File: editorial/chap04/answer_B17.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>
using namespace std;
 
int N, H[100009];
int dp[100009];
vector<int> Answer;
 
int main() {
	// 入力
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> H[i];
 
	// 動的計画法
	dp[1] = 0;
	dp[2] = abs(H[1] - H[2]);
	for (int i = 3; i <= N; i++) {
		dp[i] = min(dp[i - 1] + abs(H[i - 1] - H[i]), dp[i - 2] + abs(H[i - 2] - H[i]));
	}
 
	// 動的計画法の復元
	int Place = N;
	while (true) {
		Answer.push_back(Place);
		if (Place == 1) break;
 
		// どちらに移動するかを求める
		if (dp[Place - 1] + abs(H[Place - 1] - H[Place]) == dp[Place]) Place = Place - 1;
		else Place = Place - 2;
	}
	reverse(Answer.begin(), Answer.end());
 
	// 答えを求める
	cout << Answer.size() << endl;
	for (int i = 0; i < Answer.size(); i++) {
		if (i) cout << " ";
		cout << Answer[i];
	}
	cout << endl;
	return 0;
}

File: editorial/chap04/python/answer_B17.py
# 入力
N = int(input())
H = list(map(int, input().split()))

# 配列の準備
dp = [ None ] * N

# 動的計画法
dp[0] = 0
dp[1] = abs(H[0] - H[1])
for i in range(2, N):
	dp[i] = min(dp[i-1] + abs(H[i-1]-H[i]), dp[i-2] + abs(H[i-2]-H[i]))

# 答えの復元
# 変数 Place は現在位置（ゴールから進んでいく）
Answer = []
Place = N-1
while True:
	Answer.append(Place + 1)
	if Place == 0:
		break

	# どこから部屋 Place に向かうのが最適かを求める
	if dp[Place-1] + abs(H[Place-1]-H[Place]) == dp[Place]:
		Place = Place - 1
	else:
		Place = Place - 2
Answer.reverse()

# 答えを出力
Answer2 = [str(i) for i in Answer]
print(len(Answer))
print(" ".join(Answer2))
